In [5]:
import sys 
print(sys.version)
print(sys.path)

3.14.2 (tags/v3.14.2:df79316, Dec  5 2025, 17:18:21) [MSC v.1944 64 bit (AMD64)]
['C:\\Users\\orynb\\AppData\\Local\\Programs\\Python\\Python314\\python314.zip', 'C:\\Users\\orynb\\AppData\\Local\\Programs\\Python\\Python314\\DLLs', 'C:\\Users\\orynb\\AppData\\Local\\Programs\\Python\\Python314\\Lib', 'C:\\Users\\orynb\\AppData\\Local\\Programs\\Python\\Python314', 'C:\\Users\\orynb\\Desktop\\Tomorrow\\Piscine_AI\\modules\\ex00', '', 'C:\\Users\\orynb\\Desktop\\Tomorrow\\Piscine_AI\\modules\\ex00\\Lib\\site-packages']


## Exercise 1: Import mechanics

In [2]:
import math
from math import sqrt, pi
import math as m
from math import *     # why is this discouraged?

In [8]:
print("Form 1:", math.pi)
print("Form 2:", pi)
print("Form 3:", m.pi)
print("Form 4:", pi)

Form 1: 3.141592653589793
Form 2: 3.141592653589793
Form 3: 3.141592653589793
Form 4: 3.141592653589793


In [18]:
from greet import hello
print(hello("Astana"))

Hello, Astana!


This prevents code from accidentally running when you just want to import its functions!

In [19]:
import importlib
import greet

importlib.reload(greet)

from greet import goodbye
print(greet.hello("Astana"))
print(goodbye("Astana"))

Hello, Astana!
GoodBye, Astana!


## Exercise 2: Standard library

In [30]:
from collections import Counter

s = "the quick brown fox jumps over the lazy dog"

s_clean = s.replace(" ", "")

counts = Counter(s_clean)

tops = counts.most_common(5)

print(tops)

[('o', 4), ('e', 3), ('t', 2), ('h', 2), ('u', 2)]


In [27]:
import itertools 

teams = ["Alpha", "Beta", "Gamma", "Delta", "Epsilon"]

pairs = list(itertools.combinations(teams, 2))

print(f"{len(pairs)} matches total")

for match in pairs[:3]:
    print(f"{match[0]} vs {match[1]}")

10 matches total
Alpha vs Beta
Alpha vs Gamma
Alpha vs Delta


In [34]:
from datetime import datetime, timedelta

def days_until(date_str):
    target_date = datetime.strptime(date_str, "%Y-%m-%d").date()
    today = datetime.now().date()
    delta = target_date - today
    return delta.days

future_date_obj = datetime.now() + timedelta(days=30)
future_date_str = future_date_obj.strftime("%Y-%m-%d")

print(f"Testing with date: {future_date_str}")
print(f"Days until: {days_until(future_date_str)}")

Testing with date: 2026-07-21
Days until: 30


In [35]:
import random

random.seed(42)

random_list = [random.randint(1, 100) for _ in range(10)]

random.shuffle(random_list)

print(random_list[:5])

[95, 4, 36, 14, 32]


## Exercise 3: Writing your own module

In [45]:
import importlib
import stats
importlib.reload(stats)
from stats import mean
from stats import median
from stats import variance

data = [4, 8, 15, 16, 23, 42]
print(mean(data))
print(median(data))
print(round(variance(data), 4))

18.0
15.5
151.6667


In [47]:
import importlib
import stats

importlib.reload(stats)

print(stats.__doc__)


This module provides basic statistical operations including mean, 
median, and population variance without using external libraries.
Author: kaorynbek



## Exercise 4: Packages

In [1]:
from toolkit import word_count, clamp

print(word_count("the quick brown fox"))
print(clamp(150, 0, 100))
print(clamp(-5, 0, 100))

4
100
0


In [2]:
import toolkit.text
import toolkit.numbers

print(toolkit.text.word_count("the quick brown fox"))
print(toolkit.numbers.clamp(150, 0, 100))

4
100


## Exercise 5: Virtual environment management

I created ex05 directory and created there new virtual envirenmonts and then first i wrote: 

py -3.14 -m venv venv_project
venv_project\Scripts\Activate.ps1
pip install requests==2.31.0 rich==13.7.0
pip freeze > requirements.txt
cat requirements.txt

and then i left this venv and created new one 

py -3.14 -m venv venv_repro
venv_repro\Scripts\Activate.ps1
pip install -r requirements.txt

### And checked version with this command: 
python -c "from importlib.metadata import version; print(version('rich'))"

ML pipelines rely heavily on exact math implementations, data splits, and structural logic inside libraries.
If an underlying package updates, a function might handle missing data differently or alter matrix shapes.
This can cause a model to output completely different predictions or break down entirely. 
Pinning guarantees consistency across your laptop, your team's machines, and production servers.